src/arch/arm/64/kernel/vspace.c

```diff
@@ -647,290 +624,80 @@ exception_t checkValidIPCBuffer(vptr_t vptr, cap_t cap)
-    bool_t nonexecutable = vm_attributes_get_armExecuteNever(attributes);
-
-    if (vm_attributes_get_armPageCacheable(attributes)) {
-        return pde_pde_large_new(
-                   nonexecutable,              /* unprivileged execute never */
-                   paddr,
-#ifdef CONFIG_ARM_HYPERVISOR_SUPPORT
-                   0,
-#else
-                   1,                          /* not global */
-#endif
-                   1,                          /* access flag */
-                   SMP_TERNARY(SMP_SHARE, 0),          /* Inner-shareable if SMP enabled, otherwise unshared */
-                   APFromVMRights(vm_rights),
-#ifdef CONFIG_ARM_HYPERVISOR_SUPPORT
-                   S2_NORMAL
-#else
-                   NORMAL
-#endif
-               );
-    } else {
-        return pde_pde_large_new(
-                   nonexecutable,              /* unprivileged execute never */
-                   paddr,
-#ifdef CONFIG_ARM_HYPERVISOR_SUPPORT
-                   0,
-#else
-                   1,                          /* not global */
-#endif
-                   1,                          /* access flag */
-                   0,                          /* Ignored - Outter shareable */
-                   APFromVMRights(vm_rights),
-#ifdef CONFIG_ARM_HYPERVISOR_SUPPORT
-                   S2_DEVICE_nGnRnE
-#else
-                   DEVICE_nGnRnE
-#endif
-               );
-    }
-}
-
-static pude_t makeUser1stLevel(paddr_t paddr, vm_rights_t vm_rights, vm_attributes_t attributes)
-{
-    bool_t nonexecutable = vm_attributes_get_armExecuteNever(attributes);
-
-    if (vm_attributes_get_armPageCacheable(attributes)) {
-        return pude_pude_1g_new(
-                   nonexecutable,              /* unprivileged execute never */
-                   paddr,
-#ifdef CONFIG_ARM_HYPERVISOR_SUPPORT
-                   0,
-#else
-                   1,                          /* not global */
-#endif
-                   1,                          /* access flag */
-                   SMP_TERNARY(SMP_SHARE, 0),          /* Inner-shareable if SMP enabled, otherwise unshared */
-                   APFromVMRights(vm_rights),
-#ifdef CONFIG_ARM_HYPERVISOR_SUPPORT
-                   S2_NORMAL
-#else
-                   NORMAL
-#endif
-               );
-    } else {
-        return pude_pude_1g_new(
-                   nonexecutable,              /* unprivileged execute never */
-                   paddr,
-#ifdef CONFIG_ARM_HYPERVISOR_SUPPORT
-                   0,
-#else
-                   1,                          /* not global */
-#endif
-                   1,                          /* access flag */
-                   0,                          /* Ignored - Outter shareable */
-                   APFromVMRights(vm_rights),
-#ifdef CONFIG_ARM_HYPERVISOR_SUPPORT
-                   S2_DEVICE_nGnRnE
-#else
-                   DEVICE_nGnRnE
-#endif
-               );
+    if (page_size == ARMSmallPage) {
+        ret.words[0] |= 0x3;
```

Comment:
So, this is the only difference between making 4K page and making 2MiB and 1GiB pages? We could use a bit comment here?

Reply:
Yeah, you're right, we should document that.